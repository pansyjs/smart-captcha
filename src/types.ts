import type { CSSProperties } from 'react';

export interface SmartCaptchaResult {
  sessionId: string;
  sig: string;
}

export class SmartCaptcha {
  constructor() {

  }
}

export interface SmartCaptchaData extends SmartCaptchaResult {
  token: string;
  appKey: string;
  scene: string;
}

export interface SmartCaptchaProps {
  /** 节点类名 */
  className?: string;
  /** 节点样式 */
  style?: CSSProperties;
  /**
   * 渲染的目标元素ID
   * @default 'smart-captcha'
   */
  elementId?: string;
  /** 组件宽度 */
  width?: number;
  /** 组件高度 */
  height?: number;
  /**
   * 验证通过时触发
   */
  onSuccess?: (data: SmartCaptchaData) => void;
  /**
   * 验证失败时触发
   */
  onFailed?: () => void;
  /**
   * 验证通过/失败都会触发
   */
  onChange?: (data?: SmartCaptchaData | undefined) => void;
}

export interface SmartCaptchaRef {
  reset: () => void;
}
