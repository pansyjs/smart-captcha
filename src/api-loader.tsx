import React from 'react';
import { useExternal } from '@pansy/react-hooks';

export const ApiLoader: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const status = useExternal('//g.alicdn.com/AWSC/AWSC/awsc.js');

  return (
    <>
      {(status === 'ready') && children}
    </>
  )
}
