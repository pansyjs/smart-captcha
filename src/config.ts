export const Option = {
  appkey: 'FFFF0N00000000008D02',
  scene: 'ic_login',
};
