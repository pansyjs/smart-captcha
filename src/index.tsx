export { default as SliderCaptcha } from './components/Slider';
export { default as SmartCaptcha } from './components/Smart';

export type {
  SliderCaptchaProps,
  SmartCaptchaData,
  SmartCaptchaProps,
  SmartCaptchaRef,
} from './types';
