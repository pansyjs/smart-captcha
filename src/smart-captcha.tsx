import React, {
  forwardRef,
  useRef,
  useEffect,
  useImperativeHandle,
} from 'react';
import { Option } from './config';

import type { SmartCaptchaProps, SmartCaptchaResult, SmartCaptchaRef, } from './types';

const SmartCaptcha: React.ForwardRefRenderFunction<SmartCaptchaRef, SmartCaptchaProps> = (
  {
    className,
    style,
    elementId = 'smart-captcha',
    width = 300,
    height = 42,
    onSuccess,
    onChange,
    onFailed,
  },
  ref,
) => {
  const ic = useRef<any>();

  useImperativeHandle(ref, () => ({
    reset: () => {
      ic?.current?.reset();
    },
  }));

  useEffect(() => {
    if (window.AWSC) {
      window.AWSC.use('ic', function (state: string, module: any) {
        ic.current = module.init({
          // 声明智能验证需要渲染的目标元素ID。
          renderTo: `#${elementId}`,
          // 智能验证组件的宽度。
          width: width,
          // 智能验证组件的高度。
          height: height,
          // 智能验证组件初始状态文案。
          default_txt: '点击按钮开始智能验证',
          // 智能验证组件验证通过状态文案。
          success_txt: '验证成功',
          // 智能验证组件验证失败（拦截）状态文案。
          fail_txt: '验证失败，请在此点击按钮刷新',
          // 智能验证组件验证中状态文案。
          scaning_txt: '智能检测中',
          // 应用类型标识
          appkey: Option.appkey,
          // 使用场景标识
          scene: Option.scene,
          success: function(data: SmartCaptchaResult) {
            const value = {
              sig: data.sig,
              token: data.token,
              sessionId: data.sessionId,
              appKey: Option.appkey,
              scene: Option.scene,
            }
            onChange?.(value);
            onSuccess?.(value);
          },
          fail: function() {
            ic?.current?.reset();
            onFailed?.();
            onChange?.(undefined);
          },
          error: function() {
            ic?.current?.reset();
            onFailed?.();
            onChange?.(undefined);
          },
        })
      })
    }

    return () => {
      ic.current = undefined;
    }
  }, []);

  return (
    <div className={className} style={style} id={elementId} />
  );
};

export default forwardRef(SmartCaptcha);
