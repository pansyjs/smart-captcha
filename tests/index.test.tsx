import * as React from 'react';
import { act, render, fireEvent, } from '@testing-library/react';
import { SmartCaptcha } from '../src';

describe('SmartCaptcha', () => {
  it('should be defined', () => {
    expect(SmartCaptcha).toBeDefined();
  });

  it('renders correctly', async () => {
    const { container } = render(<SmartCaptcha />);

    const script = document.querySelector('script') as HTMLScriptElement;

    act(() => {
      fireEvent.load(script);
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('mount correctly', () => {
    expect(() => render(<SmartCaptcha />)).not.toThrow();
  });
});
