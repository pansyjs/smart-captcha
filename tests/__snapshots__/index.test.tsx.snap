// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmartCaptcha renders correctly 1`] = `
<div
  id="smart-captcha"
/>
`;
